### PR TITLE
Add Teacher Dashboard tab

### PR DESCRIPTION
**Pull request by denys-zarubin**
@auraz @polesye  @scherbatey Please, review

### CODE REVIEW

File: lms/djangoapps/teacher_dashboard/tabs.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+    """
+
+    type = "teacher_dashboard"
+    title = ugettext_noop("Teacher Dashboard")
+    view_name = "dashboard_view_handler"
+    is_dynamic = True
+
+    @classmethod
+    def is_enabled(cls, course, user=None):
+        return True
```
Comment (polesye) [inline lms/djangoapps/teacher_dashboard/tabs.py]: Create a FEATURE flag [here](https://github.com/Livit/Livit.Learn.EdX/blob/labster/teacher_dashboard_tab/lms/envs/labster.py#L11) and use it instead `True`

File: lms/djangoapps/teacher_dashboard/urls.py [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+
+urlpatterns = patterns(
+        '',
+        url(r'^teacher_dashboard/?$',
+            'teacher_dashboard.views.dashboard_view', name='dashboard_view_handler'),
+        url(r'^licenses/?$', 'teacher_dashboard.views.licenses_api_call'),
+        url(r'^licenses/(?P<license_pk>[a-zA-Z0-9\-]+)/simulations/?$', 'teacher_dashboard.views.simulations_api_call'),
+        url(r'^licenses/(?P<license_pk>[a-zA-Z0-9\-]+)/simulations/(?P<simulation_pk>[a-zA-Z0-9]+)/students/?$',
+            'teacher_dashboard.views.students_api_call'),
+
```
Comment (polesye) [inline lms/djangoapps/teacher_dashboard/urls.py]: remove empty line.

File: lms/djangoapps/teacher_dashboard/views.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+
+def licenses_api_call(request):
+    """
+    Redirect Api call to Labster API
+    """
+
+    headers = setup_headers()
+    url = urlparse.urljoin(settings.LABSTER_API_URL, settings.LABSTER_API_URLS.get('licenses'))
+    response = requests.get(url, data=dict(request.GET), headers=headers)
```
Comment (polesye) [inline lms/djangoapps/teacher_dashboard/views.py]: I'd like to have better error handling. See https://github.com/Livit/Livit.Learn.EdX/blob/labster/teacher_dashboard_tab/lms/djangoapps/labster_course_license/views.py#L96

File: lms/djangoapps/teacher_dashboard/views.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    }
+    return render_to_response('teacher_dashboard/dashboard.html', context)
+
+
+def licenses_api_call(request):
+    """
+    Redirect Api call to Labster API
+    """
+
+    headers = setup_headers()
```
Comment (polesye) [inline lms/djangoapps/teacher_dashboard/views.py]: `setup_headers` looks redundant.

File: lms/envs/labster.py [hunk 5]
```diff
@@ -21,7 +21,7 @@
 LABSTER_WIKI_LINK = LABSTER_SETTINGS.get('LABSTER_WIKI_LINK', 'https://theory.labster.com/')
 LABSTER_API_AUTH_TOKEN = LABSTER_AUTH.get('LABSTER_API_AUTH_TOKEN', '')
 
-LABSTER_API_URL = LABSTER_SETTINGS.get('LABSTER_API_URL', '')
+LABSTER_API_URL = LABSTER_SETTINGS.get('LABSTER_API_URL', 'http://staging-api2.labster.com')
```
Comment (polesye) [inline lms/envs/labster.py]: Add default value in your local `lms.env.json`.

File: lms/envs/labster.py [hunk 6]
```diff
@@ -32,3 +32,10 @@
     LABSTER_ENDPOINTS[endpoint] = value
 
 LABSTER_DEFAULT_LTI_ID = LABSTER_SETTINGS.get('LABSTER_DEFAULT_LTI_ID', 'MC')
+LABSTER_API_URLS = {
+    "coaches": "/labster/api/coaches/",
```
Comment (polesye) [inline lms/envs/labster.py]: We already have LABSTER_ENDPOINTS for these purposes

Comment (polesye) [inline lms/envs/labster.py]: I believe that we should have possibility to manage them via json file.

File: lms/static/teacher_dashboard/css/main.css [hunk 7]
```diff
@@ -0,0 +1,221 @@
+/*html {*/
+  /*font-size: 10px;*/
+/*}*/
+
+/*body {*/
+  /*padding-top: 50px;*/
+  /*padding-bottom: 20px;*/
+/*}*/
```
Comment (polesye) [inline lms/static/teacher_dashboard/css/main.css]: remove commented block.

File: lms/static/teacher_dashboard/js/templates/coach_selector.underscore [hunk 8]
```diff
@@ -0,0 +1,9 @@
+<label class="col-sm-2 control-label" for="coaches-select">Coaches:</label>
+<div class="col-sm-10">
```
Comment (polesye) [inline lms/static/teacher_dashboard/js/templates/coach_selector.underscore]: Do not need this template in edx.

File: lms/static/teacher_dashboard/css/main.css [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+.loading-bar {
+  text-align: center;
+  font-size: 1.1em;
+}
+
+.loading-bar span {
+  margin-right: 5px;
+}
+
+#coaches {
```
Comment (polesye) [inline lms/static/teacher_dashboard/css/main.css]: Do not need this rule in edx.

File: lms/envs/labster.py (hunk 6 above)
Comment (polesye) [inline lms/envs/labster.py]: Do not need this endpoint `coaches` in edx.

File: lms/static/teacher_dashboard/js/app/collections/coach.js [hunk 10]
```diff
@@ -0,0 +1,12 @@
+define(["backbone", "app/models/coach"], function(Backbone, CoachModel) {
+  var CoachCollection = Backbone.Collection.extend({
```
Comment (polesye) [inline lms/static/teacher_dashboard/js/app/collections/coach.js]: Do not need this collection in edx.

File: lms/static/teacher_dashboard/js/app/views/coach_selector.js [hunk 11]
```diff
@@ -0,0 +1,21 @@
+define([
```
Comment (polesye) [inline lms/static/teacher_dashboard/js/app/views/coach_selector.js]: Do not need this view in edx.

File: lms/static/teacher_dashboard/js/main.js [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+    "domReady", "jquery", "app/collections/license", "app/views/license_list", "app/collections/coach",
+    "app/views/coach_selector", "app/utils", "bootstrap"
+], function (domReady, $, LicenseCollection, LicenseListView, CoachCollection, CoachSelectorView, utils) {
+    domReady(function () {
+        var buildLicenseList = _.once(function (collection) {
+            var view = new LicenseListView({collection: collection});
+            $(document.getElementById("content1")).html(view.$el);
+
+        });
+        $coach = $(document.getElementById("coach")).data("metadata");
```
Comment (polesye) [inline lms/static/teacher_dashboard/js/main.js]: Remove this line

File: lms/static/teacher_dashboard/js/main.js [hunk 13]
```diff
@@ -0,0 +1,24 @@
+require([
+    "domReady", "jquery", "app/collections/license", "app/views/license_list", "app/collections/coach",
```
Comment (polesye) [inline lms/static/teacher_dashboard/js/main.js]: remove `app/collections/coach` and `app/views/coach_selector`

File: lms/static/teacher_dashboard/js/main.js [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+    "app/views/coach_selector", "app/utils", "bootstrap"
+], function (domReady, $, LicenseCollection, LicenseListView, CoachCollection, CoachSelectorView, utils) {
```
Comment (polesye) [inline lms/static/teacher_dashboard/js/main.js]: remove `CoachSelectorView`, `CoachCollection`

File: lms/templates/teacher_dashboard/teacher_dashboard.html [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+  if is_active:
+    return "active"
+  return ""
+
+def selected(is_selected):
+  return "selected" if is_selected else ""
+
+%>
+
+
```
Comment (polesye) [inline lms/templates/teacher_dashboard/teacher_dashboard.html]: remove empy lines

File: lms/templates/teacher_dashboard/dashboard.html [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+    <div class="info-wrapper">
+        <section class="updates">
+            <div id="coach" data-metadata='${coach}' style="display: none">
+            </div>
+            <!-- Example row of columns -->
+            <div id="content1"></div>
+            <hr>
+        </section>
+    </div>
+</section>
```
Comment (polesye) [inline lms/templates/teacher_dashboard/dashboard.html]: add empty line.

File: lms/templates/teacher_dashboard/dashboard.html [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+<%include file="/teacher_dashboard/teacher_dashboard.html" args="active_page='teacher_dashboard'" />
+
+
+<section class="container">
+    <div class="info-wrapper">
+        <section class="updates">
+            <div id="coach" data-metadata='${coach}' style="display: none">
+            </div>
+            <!-- Example row of columns -->
+            <div id="content1"></div>
```
Comment (polesye) [inline lms/templates/teacher_dashboard/dashboard.html]: bad id `content1`

File: lms/urls.py [hunk 18]
```diff
@@ -724,6 +724,15 @@
     url(r'^courses/{}/'.format(settings.COURSE_ID_PATTERN), include('labster_course_license.urls')),
 )
 
+#Teacher Dashboard Tab
+urlpatterns += (
+    url(r'^courses/{}/'.format(settings.COURSE_ID_PATTERN), include('teacher_dashboard.urls')),
+)
+
+urlpatterns += (
```
Comment (polesye) [inline lms/urls.py]: I believe that we can use one include.

File: lms/templates/teacher_dashboard/teacher_dashboard.html [hunk 19]
```diff
@@ -0,0 +1,59 @@
+## mako
```
Comment (polesye) [inline lms/templates/teacher_dashboard/teacher_dashboard.html]: Why do we need this file? Can we avoid it?

File: lms/templates/teacher_dashboard/dashboard.html [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+
+
+
+<%include file="/teacher_dashboard/teacher_dashboard.html" args="active_page='teacher_dashboard'" />
+
+
+<section class="container">
+    <div class="info-wrapper">
+        <section class="updates">
+            <div id="coach" data-metadata='${coach}' style="display: none">
```
Comment (polesye) [inline lms/templates/teacher_dashboard/dashboard.html]: Redundant.

You can use: 

``` html
<script>window.Labster = ${coach};</script>
```

File: lms/static/teacher_dashboard/js/main.js [hunk 21]
```diff
@@ -0,0 +1,24 @@
...
+        window.Labster.coach = $coach;
+        buildLicenseList(licenseCollection);
+        if (coach) {
+            licenseCollection.url = utils.getUrl("licenses", {}, true);
+            licenseCollection.fetch();
+        } else {
+            licenseCollection.reset();
+        }
+
+
```
Comment (polesye) [inline lms/static/teacher_dashboard/js/main.js]: remove empty lines.

File: lms/static/teacher_dashboard/js/main.js [hunk 22]
```diff
@@ -0,0 +1,24 @@
...
+        });
+        $coach = $(document.getElementById("coach")).data("metadata");
+        var licenseCollection = LicenseCollection.factory();
+        window.Labster.coach = $coach;
+        buildLicenseList(licenseCollection);
+        if (coach) {
+            licenseCollection.url = utils.getUrl("licenses", {}, true);
+            licenseCollection.fetch();
+        } else {
```
Comment (polesye) [inline lms/static/teacher_dashboard/js/main.js]: Is this `else` possible?

File: lms/static/teacher_dashboard/js/main.js [hunk 23]
```diff
@@ -0,0 +1,24 @@
+require([
+    "domReady", "jquery", "app/collections/license", "app/views/license_list", "app/collections/coach",
+    "app/views/coach_selector", "app/utils", "bootstrap"
+], function (domReady, $, LicenseCollection, LicenseListView, CoachCollection, CoachSelectorView, utils) {
+    domReady(function () {
+        var buildLicenseList = _.once(function (collection) {
```
Comment (polesye) [inline lms/static/teacher_dashboard/js/main.js]: do not need `_.once` here and function `buildLicenseList` looks redundant.

File: lms/static/teacher_dashboard/js/main.js [hunk 24]
```diff
@@ -0,0 +1,24 @@
...
+            var view = new LicenseListView({collection: collection});
+            $(document.getElementById("content1")).html(view.$el);
+
+        });
+        $coach = $(document.getElementById("coach")).data("metadata");
+        var licenseCollection = LicenseCollection.factory();
+        window.Labster.coach = $coach;
```
Comment (polesye) [inline lms/static/teacher_dashboard/js/main.js]: remove this line.

File: lms/static/teacher_dashboard/js/main.js [hunk 25]
```diff
@@ -0,0 +1,24 @@
+require([
+    "domReady", "jquery", "app/collections/license", "app/views/license_list", "app/collections/coach",
+    "app/views/coach_selector", "app/utils", "bootstrap"
+], function (domReady, $, LicenseCollection, LicenseListView, CoachCollection, CoachSelectorView, utils) {
+    domReady(function () {
+        var buildLicenseList = _.once(function (collection) {
+            var view = new LicenseListView({collection: collection});
+            $(document.getElementById("content1")).html(view.$el);
```
Comment (polesye) [inline lms/static/teacher_dashboard/js/main.js]: `content1` is a bad name.

File: lms/djangoapps/teacher_dashboard/utils.py [hunk 26]
```diff
@@ -0,0 +1,9 @@
+from django.conf import settings
```
Comment (polesye) [inline lms/djangoapps/teacher_dashboard/utils.py:4]: I think it is redundant module.

File: lms/templates/teacher_dashboard/dashboard.html (hunk 20 above)
Comment (polesye) [inline lms/templates/teacher_dashboard/dashboard.html]: **But it is much better to get the user from request**

File: lms/static/teacher_dashboard/js/app/utils.js [hunk 27]
```diff
@@ -0,0 +1,60 @@
...
+      URL_BASE = "/teacher/api/v0",
+      URL_TEMPLATES;
+
+  URL_TEMPLATES = {
+    "licenses": _.template("<%= base %>/licenses/"),
+    "simulations": _.template("<%= base %>/licenses/<%=license_id %>/simulations/"),
+    "students": _.template("<%= base %>/licenses/<%=license_id %>/simulations/<%=simulation_id %>/students/")
+  };
+
+  utils.addUserToUri = function(uri) {
```
Comment (polesye) [inline lms/static/teacher_dashboard/js/app/utils.js]: You don't need this method because you can get the user from template.

File: lms/static/teacher_dashboard/js/app/utils.js [hunk 28]
```diff
@@ -0,0 +1,60 @@
...
+    var url = "";
+
+    if (!URL_TEMPLATES[type]) {
+      throw new Error("Url with type `" + type + "` does not exist.");
+    }
+    context = _.extendOwn({base: URL_BASE}, context);
+    url = URL_TEMPLATES[type](context);
+
+    if (addUser) {
+      return utils.addUserToUri(url);
```
Comment (polesye) [inline lms/static/teacher_dashboard/js/app/utils.js]: You don't need this because you can get the user from template.

File: lms/djangoapps/teacher_dashboard/views.py [hunk 29]
```diff
@@ -0,0 +1,62 @@
...
+    # Course is needed for display others tabs
+    course_key = SlashSeparatedCourseKey.from_deprecated_string(course_id)
+    with modulestore().bulk_operations(course_key):
+        course = get_course_by_id(course_key, depth=2)
+    coach = request.user.email
+    context = {
+        'request': request,
+        'course_id': course_id,
+        'cache': None,
+        'coach': coach,
```
Comment (polesye) [inline lms/djangoapps/teacher_dashboard/views.py]: You don't need this because you can get the user from template.

File: lms/djangoapps/teacher_dashboard/tabs.py (hunk 1 above)
Comment (polesye) [inline lms/djangoapps/teacher_dashboard/tabs.py]: **It must to be shown only for Instructors.**

**Comment by polesye** [comment]
More points:
- avoid duplication of js dependecies (require js, jquery, URI etc).
- use consumer key to get a list of licenses for the course.
- avoid of using bootstrap in edx and try to use http://ux.edx.org/ .
- add i18n.

File: lms/djangoapps/teacher_dashboard/views.py [hunk 30]
```diff
@@ -0,0 +1,62 @@
...
+    Teacher dashboard renders data using backbone from /static/teacher_dashboard/js
+    """
+    # Course is needed for display others tabs
+    course_key = SlashSeparatedCourseKey.from_deprecated_string(course_id)
+    with modulestore().bulk_operations(course_key):
+        course = get_course_by_id(course_key, depth=2)
+    coach = request.user.email
+    context = {
+        'request': request,
+        'course_id': course_id,
```
Comment (polesye) [inline lms/djangoapps/teacher_dashboard/views.py]: Why do we need this line?

File: lms/djangoapps/teacher_dashboard/views.py [hunk 31]
```diff
@@ -0,0 +1,62 @@
...
+
+
+@login_required
+def dashboard_view(request, course_id):
+    """
+    Teacher dashboard renders data using backbone from /static/teacher_dashboard/js
+    """
+    # Course is needed for display others tabs
+    course_key = SlashSeparatedCourseKey.from_deprecated_string(course_id)
+    with modulestore().bulk_operations(course_key):
```
Comment (polesye) [inline lms/djangoapps/teacher_dashboard/views.py:33]: Why do we need this code?

File: lms/djangoapps/teacher_dashboard/utils.py (hunk 26 above)
Comment (denys-zarubin) [inline lms/djangoapps/teacher_dashboard/utils.py:4]: > I think it is redundant module.

We are taking from settings LABSTER_API_AUTH_TOKEN. I don't think that settings are reudndant

File: lms/djangoapps/teacher_dashboard/views.py (hunk 31 above)
Comment (denys-zarubin) [inline lms/djangoapps/teacher_dashboard/views.py:33]: > Why do we need this code?

We are using course for displaying tabs.

File: lms/djangoapps/teacher_dashboard/utils.py (hunk 26 above)
Comment (polesye) [inline lms/djangoapps/teacher_dashboard/utils.py:4]: > We are taking from settings LABSTER_API_AUTH_TOKEN. I don't think that settings are reudndant

Yes, after moving `_send_request` in this file it is important to have it there.

Comment (denys-zarubin) [inline lms/djangoapps/teacher_dashboard/utils.py:4]: > Yes, after moving _send_request in this file it is important to have it there.

We will ignore this

**Comment by polesye** [comment]
@den99f do not do:
- avoid of using bootstrap in edx and try to use http://ux.edx.org/ .
- add i18n.

File: openedx/core/djangoapps/labster/course/utils.py [hunk 32]
```diff
@@ -1,21 +1,44 @@
...
 
-from django.contrib.auth.models import User
-from contentstore.utils import add_instructor, remove_all_instructors
-
+from openedx.core.djangoapps.labster.exceptions import LtiPassportError
 
 log = logging.getLogger(__name__)
 
-
-def set_staff(course_key, emails):
```
Comment (polesye) [inline openedx/core/djangoapps/labster/course/utils.py:11]: @den99f Why it was removed?

File: openedx/core/djangoapps/labster/course/views.py [hunk 33]
```diff
@@ -2,6 +2,7 @@
 import logging
 
 from django.http import Http404
+from django.contrib.auth.models import User
```
Comment (polesye) [inline openedx/core/djangoapps/labster/course/views.py:5]: I think we have to move into a separate PR.

File: lms/urls.py [hunk 34]
```diff
@@ -724,6 +724,12 @@
     url(r'^courses/{}/'.format(settings.COURSE_ID_PATTERN), include('labster_course_license.urls')),
 )
 
+#Teacher Dashboard Tab
+urlpatterns += (
+    url(r'^courses/{}/'.format(settings.COURSE_ID_PATTERN), include('teacher_dashboard.urls')),
```
Comment (polesye) [inline lms/urls.py:729]: Do we still need this endpoint?

**Comment by polesye** [comment]
@den99f Few small remarks, verify that everything works well. :+1:

File: lms/static/teacher_dashboard/css/main.css [hunk 35]
```diff
@@ -0,0 +1,221 @@
...
+  display: block;
+  font-size: 1.3rem;
+  line-height: 1.3;
+}
+
+.license-item {
+  border: 1px solid rgba(0, 0, 0, 1);
+  margin-bottom: 20px;
+  line-height: 2.5rem;
+}
```
Comment (polesye) [inline lms/static/teacher_dashboard/css/main.css:39]:
``` css
.license-item {
   border: 1px solid rgba(0, 0, 0, 1);
   line-height: 2.5rem;
 }
.license-item + .license-item {
   margin-top: 20px;
}
```

File: lms/djangoapps/teacher_dashboard/urls.py (hunk 2 above)
Comment (denys-zarubin) [inline lms/djangoapps/teacher_dashboard/urls.py]: Done. Also change style of url due to edx pylint

File: lms/urls.py (hunk 34 above)
Comment (denys-zarubin) [inline lms/urls.py:729]: Yes, we need
